src/WebJobs.Script.WebHost/Management/InstanceManager.cs

```diff
@@ -44,51 +44,62 @@ public bool StartAssignment(HostAssignmentContext assignmentContext)
                 {
                     if (_assignmentContext != null)
                     {
-                        return _assignmentContext.Equals(assignmentContext);
+                        return _assignmentContext.Equals(context);
                     }
-                    _assignmentContext = assignmentContext;
+                    _assignmentContext = context;
                 }
 
-                // Queue the task in background, and return to caller immediately
-                Task.Run(() => Specialize(assignmentContext));
+                // start the specialization process in the background
+                Task.Run(async () => await Specialize(context));
 
                 return true;
             }
             else
             {
                 // No lock needed here since _assignmentContext is not null when we are here
-                return _assignmentContext.Equals(assignmentContext);
+                return _assignmentContext.Equals(context);
             }
         }
 
         private async Task Specialize(HostAssignmentContext assignmentContext)
         {
             try
             {
-                var zip = assignmentContext.ZipUrl;
-                if (!string.IsNullOrEmpty(zip))
-                {
-                    // download zip and extract
-                    var filePath = Path.GetTempFileName();
-
-                    await DownloadAsync(new Uri(zip), filePath);
+                // set a flag which will cause any incoming http requests to buffer
+                // until specialization is complete
+                WebScriptHostManager.DelayRequests = true;
```

Comment:
Request buffering is controlled via this flag which is set to true during specialization.